### PR TITLE
Hero and HeroRole updated with data and  managment command

### PR DESCRIPTION
**Pull request by Ayub-Khan**
Wrote Managment Command to update Hero and HeroRole from data and added data file."

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,6 +12,7 @@ help:
 	@echo "  serve                      serve Dota World project at 0.0.0.0:4444"
 	@echo "  static                     build and compress static assets"
 	@echo "  clean_static               delete compiled/compressed static assets"
+	@echo "  update_hero_models         delete old data and create new models from datafile"
```
Comment (zubair-arbi) [inline Makefile]: We can make the command name short `update_heros`

File: heros/management/commands/update_hero_models.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from django.core.management.base import BaseCommand
+from heros.models import Hero, HeroRole
+from django.contrib.staticfiles import finders
+
+
+class Command(BaseCommand):
+    help = 'Updates the Hero and Hero Role Models with data'
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: `Update the Hero app models from preformated data files`

File: heros/management/commands/update_hero_models.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    def handle(self, *args, **options):
+
+        self.stdout.write('Starting to Update Hero and Hero Role Models.')
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: We can modularize this handle method like:
```python


def handle(self, *args, **options):
    self.stdout.write('Finding prefomated Hero data files...')
    data_file = finders.find('data/hero_data.txt')
    if not data_file:
        raise CommandError('Unable to file prefomated Hero data file "hero_data.txt".')

    self.stdout.write('Heros data file found. Starting to Update Hero models.')
    self.update_hero_models(data_file)
    self.stdout.write('Hero models have been updated.')

    self.stdout.write('Starting to update HeroRole models.')
    self.update_hero_role_models(data_file)
    self.stdout.write('HeroRole models have been updated.')

    self.stdout.write('Command completed successfully.')
```

File: heros/models.py [hunk 4]
```diff
@@ -9,13 +9,13 @@
 class Hero(TimeStampedModel):
 
     HERO_CATEGORY_CHOICES = (
-        ('strength', 'Strength hero'),
-        ('agility', 'Agility hero'),
-        ('intelligence', 'Intelligence hero'),
+        ('Strength', 'Strength hero'),
```
Comment (zubair-arbi) [inline heros/models.py]: please revert these three choices as the choice name should be in lower case.

File: heros/models.py [hunk 5]
```diff
@@ -9,13 +9,13 @@
...
+        ('Agility', 'Agility hero'),
+        ('Intelligence', 'Intelligence hero'),
     )
     ATTACK_TYPE_CHOICES = (
-        ('malee', 'Malee attack'),
-        ('ranged', 'Ranged attack'),
+        ('Melee', 'Melee attack'),
```
Comment (zubair-arbi) [inline heros/models.py]:
```python
('option_name_for_handling_on_backend', 'Option value to display to users'),
```

File: heros/models.py [hunk 6]
```diff
@@ -49,7 +49,6 @@ def __unicode__(self):
 class HeroRole(models.Model):
 
     ROLE_NAME_CHOICES = (
-        ('complexity', 'Complexity'),
```
Comment (zubair-arbi) [inline heros/models.py:52]: I know we are not using it but I would suggest we keep this option as its provide by dota in his game.

File: requirements.txt [hunk 7]
```diff
@@ -17,5 +17,6 @@ requests==2.18.3
 static==1.1.1
 static3==0.7.0
 whitenoise==3.3.0
+nose==1.3.7
```
Comment (zubair-arbi) [inline requirements.txt]: alphabetically order packages

File: heros/management/commands/update_hero_models.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        self.stdout.write('Starting finding data file')
+        data_file = finders.find('data/hero_data.txt')
+        self.stdout.write('Removing previous Model Data.')
+        HeroRole.objects.all().delete()
+        Hero.objects.all().delete()
+        self.stdout.write('Beginning to update Hero Model')
+        with open(data_file, 'r') as hero_data:
+            for line in hero_data:
+                unpacked_data = line.strip().split('-')
+                image_name = unpacked_data[0].lower().replace(" ", "_").replace("'", "") + "_vert.jpg"
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: Please be consistent with single quotes

Comment (zubair-arbi) [reply heros/management/commands/update_hero_models.py]: please create helper method `get_hero_icon_file_path` which will return a relative path `'/hero-icons/'+image_name`

File: heros/management/commands/update_hero_models.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+        self.stdout.write('Starting to Update Hero and Hero Role Models.')
+        self.stdout.write('Starting finding data file')
+        data_file = finders.find('data/hero_data.txt')
+        self.stdout.write('Removing previous Model Data.')
+        HeroRole.objects.all().delete()
+        Hero.objects.all().delete()
+        self.stdout.write('Beginning to update Hero Model')
+        with open(data_file, 'r') as hero_data:
+            for line in hero_data:
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: `line` -> `hero_data_line`

File: heros/management/commands/update_hero_models.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+                unpacked_data = line.strip().split('-')
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: `unpacked_hero_data`

File: heros/management/commands/update_hero_models.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    def handle(self, *args, **options):
+
+        self.stdout.write('Starting to Update Hero and Hero Role Models.')
+        self.stdout.write('Starting finding data file')
+        data_file = finders.find('data/hero_data.txt')
+        self.stdout.write('Removing previous Model Data.')
+        HeroRole.objects.all().delete()
+        Hero.objects.all().delete()
+        self.stdout.write('Beginning to update Hero Model')
+        with open(data_file, 'r') as hero_data:
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: with open(data_file, 'r') as heros_data:

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
Overall functionality looks good.

File: heros/models.py (hunk 6 above)
Comment (Ayub-Khan) [reply heros/models.py:52]: It has no use here. batter if make it as a seprate field in the HeroRole model. let those those choices only be related to roles.

File: Makefile [hunk 12]
```diff
@@ -12,6 +12,7 @@ help:
 	@echo "  serve                      serve Dota World project at 0.0.0.0:4444"
 	@echo "  static                     build and compress static assets"
 	@echo "  clean_static               delete compiled/compressed static assets"
+	@echo "  update_hero                Update the Hero app models from pre-formatted data files"
```
Comment (zubair-arbi) [inline Makefile]: nit `update_heros`

File: heros/management/commands/update_hero_models.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
+from django.core.management.base import BaseCommand, CommandError
+from heros.models import Hero, HeroRole
+from django.contrib.staticfiles import finders
+
+
+class Command(BaseCommand):
+    help = 'Update the Hero app models from pre-formatted data files'
+
+    def remove_old_models_data(self):
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: `remove_hero_models_data`

File: heros/management/commands/update_hero_models.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        HeroRole.objects.all().delete()
+        Hero.objects.all().delete()
+
+    def get_hero_icon_file_path(self, hero_name):
+        return '/hero-icons/' + hero_name.lower().replace(' ', '_').replace("'", '') + '_vert.jpg'
+
+    def update_hero_models(self, data_file):
+        with open(data_file, 'r') as hero_data:
+            for hero_data_line in hero_data:
+                unpacked_hero_data = hero_data_line.strip().split('-')
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: nit: please provide a comment specifying an example hero data line

File: heros/management/commands/update_hero_models.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+                Hero.objects.create(
+                    name=unpacked_hero_data[0],
+                    category=unpacked_hero_data[1].lower(),
+                    attack_type=unpacked_hero_data[2].lower(),
+                    photo=self.get_hero_icon_file_path(unpacked_hero_data[0])
+                )
+
+    def update_hero_role_models(self, data_file):
+        with open(data_file, 'r') as hero_data:
+            for line in hero_data:
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: nit `line` -> `hero_data_line`

File: heros/management/commands/update_hero_models.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+    def remove_old_models_data(self):
+        HeroRole.objects.all().delete()
+        Hero.objects.all().delete()
+
+    def get_hero_icon_file_path(self, hero_name):
+        return '/hero-icons/' + hero_name.lower().replace(' ', '_').replace("'", '') + '_vert.jpg'
+
+    def update_hero_models(self, data_file):
+        with open(data_file, 'r') as hero_data:
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: its all heros data, `hero_data` -> `heros_data`

File: heros/management/commands/update_hero_models.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+        self.stdout.write('Starting to Update Hero and Hero Role Models.')
+        self.stdout.write('Finding pre-formatted Hero data files...')
+        data_file = finders.find('data/hero_data.txt')
+        if not data_file:
+            raise CommandError('Unable to file pre-formatted Hero data file "hero_data.txt".')
+
+        self.stdout.write('Removing previous Model Data.')
+        self.remove_old_models_data()
+
+        self.stdout.write('Beginning to update Hero Model')
```
Comment (zubair-arbi) [inline heros/management/commands/update_hero_models.py]: `Beginning` -> `Starting` same as for `HeroRole` msg.

**Review by zubair-arbi — APPROVED** [review]
Just few small suggestion otherwise LGTM 👍 
Please make sure to have just one migration before merging this PR.

**Pull request merged**